FreeIPA/RHIdM API dynamic inventory script & plugin (Mk II)

Aethylred opened the pull request:
A dynamic inventory script that uses the FreeIPA API with HTTPS authentication. No Kerberos required. Requires the python_freeipa module.

This is a duplicate of https://github.com/ansible/ansible/pull/41570 because I forgot to branch before developing a new feature.

SUMMARY
Adds a dynamic inventory script  and inventory plugin that uses the FreeIPA/RHIdM API with HTTPS authentication. The inventory should be the same with using either the script or the plugin.

## Dynamic Inventory Script
- dynamic inventory script works when run as a user with Kerberos authentication (requires a Kerberos ticket for the user/process calling the script)
- dynamic inventory script works when run with:
-- correct command line arguments
-- correct environment variables set either from the command line or Ansible variables
-- can be cut-pasted into Ansible AWX/Tower

## Inventory Plugins
- has to be whitelisted in an `ansible.cgf` somewhere
- works with an appropriate inventory YAML file
- Kerberos version requires the user/role/process to have a valid Kerberos ticket*

\* um... I don't have a sensible method for that... I suspect it's one of the reasons the original ipalib based dynamic inventory script wasn't converted into a plugin.

## New Featured
- Behavior of the plug-in _matches_ the behavior of the inventory script.
- Inventory replicates the nested host groups from IPA, this enhancement should not interfere with previous use of the old script as the 'old' groups still exist, they just now include nested relationships.
- Allows a host group to be specified to act as the parent for the inventory, allowing the inventory to be limited to part of the IPA hosts directory (important for host licensing for Tower)

## Dependencies
- HTTPS connections requires [python_freeipa](https://pypi.org/project/python-freeipa/) and [urllib3](https://pypi.org/project/urllib3/) for HTTPS authentication
- Kerberos authentication requires  [ipalib](https://pypi.org/project/ipalib/) and [ipaclient](https://pypi.org/project/ipaclient/) to use Kerberos authentication
-- Kerberos authentication requires that the Ansible user/process environment has valid Kerberos authentication & keytab

ISSUE TYPE
Feature Pull Request
COMPONENT NAME
Dynamic Inventory Script
Inventory Plugin

ANSIBLE VERSION
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

Aethylred commented:
Tests failing due to a TLS certificate error in an area of the code not related to this PR. https://app.shippable.com/github/ansible/ansible/runs/70931/28/tests

I'm going to try and do the rebase to reduce the number of commits.

Aethylred committed:
Create a FreeIPA/RHIdM IPA inventory plugin

My first plugin
Now testing Kerberos auth with ipalib
Changing the parent class to BaseInventoryPlugin
FreeIPA/RHIdM API dynamic inventory script

A dynamic inventory script that uses the FreeIPA API with HTTPS authentication. No Kerberos required. Requires the python_freeipa module.
Debugging


Documentation

...and remove all_members
Cleaning lint

Aethylred commented:
Tests now failing due to a failure in obtaining Docker images. Is there a way of re-running shippable jobs?

mattclay commented:
CI jobs have been restarted.

mkrupcale commented:
cc @Akasurde 

This dependency on [python_freeipa](https://python-freeipa.readthedocs.io/en/latest/), which itself depends on [Python Requests](http://docs.python-requests.org/), for a FreeIPA HTTP(S) client seems unnecessary given that Ansible already has `ansible.module_utils.ipa.IPAClient`. On quick inspection, these clients do more or less the same thing[1,2]. See [`ipa_user`](https://github.com/ansible/ansible/blob/48280463f2c8e60a222ac793fafb118ad9ce3c90/lib/ansible/modules/identity/ipa/ipa_user.py#L292) module for example usage.

See also #36326 which requests Kerberos authentication for current FreeIPA modules `ipa_*` which only use HTTP(S) authentication. Perhaps this is an opportunity to address this request as well, since you appear to have done some work on the Kerberos side here.

[1] [`ansible.module_utils.ipa.IPAClient.login`](https://github.com/ansible/ansible/blob/d8bdf38b50152417d3a1dd2facdb6599467b2ade/lib/ansible/module_utils/ipa.py#L57)
[2] [`python_freeipa.Client.login`](https://github.com/opennode/python-freeipa/blob/0e0dc462655eeeeb8a5eaa7217512f1665e091c2/src/python_freeipa/client.py#L36)

Aethylred commented:
@mkrupcale I think that `ansible.module_utils.ipa.IPAClient` could be used and would eliminate the dependency. I'm coming around to look at this again soonish (we need to trim the number of nodes in our Ansible Tower inventory), but if I did that I'd drop the Kerberos support in the plugin as I don't have an environment in which I can test it.

I didn't do any of the Kerberos work, it already existed before this PR, I just replicated what was in the old inventory script into the new plugin, I'd have left it out if it wasn't important to preserve existing functionality.

jamescassell commented:
FWIW, i'm a huge fan of keeping the kerberos auth feature. Thanks for leaving it in!

Aethylred committed:
Import MutableMappings from ansible.module_utils.common._collections_compat
Pass hostgroup to IPA inventory script
pass it to kerberos sessions too
cast hostgroup as a str
that was a bad idea
but needs to be unicode for kerberos method


decode string to utf-8 not encode
grab ipahostgroup from the environment
Ah! use hostgroup to test not args.hostgroup
FreeIPA/RHIdM API dynamic inventory script

A dynamic inventory script that uses the FreeIPA API with HTTPS authentication. No Kerberos required. Requires the python_freeipa module.
resolve conflicts from rebase

resolve conflicts from rebase

Import MutableMappings from ansible.module_utils.common._collections_compat

Aethylred commented:
OK, I've looked at using `ansible.module_utils.ipa.IPAClient` however, it's targeted specifically at being executed from an Ansible module (which an inventory plugin is not)

I'd have to build a dummy `AnsibleModule` object with the `ipa_argument_spec()` to pass to the `IPAClient` initialiser which requires 5 parameters, including `module`. To work in this way correctly `IPAClient` should also be able to handle an `AnsiblePlugin` in the same 'role' as the parent object using the `IPAClient` object to get information from the IPA server.

If there was documentation or an example of `ansible.module_utils.ipa.IPAClient` being used in an `AnsiblePlugin` that I could have followed I may have made more progress.

Aethylred committed:
Pass hostgroup to IPA inventory script
pass it to kerberos sessions too
cast hostgroup as a str
that was a bad idea
but needs to be unicode for kerberos method


decode string to utf-8 not encode
grab ipahostgroup from the environment
Ah! use hostgroup to test not args.hostgroup

Aethylred committed:
FreeIPA/RHIdM API dynamic inventory script

A dynamic inventory script that uses the FreeIPA API with HTTPS authentication. No Kerberos required. Requires the python_freeipa module.
resolve conflict with freeipa inventory script

Aethylred commented:
Rebased for your consideration.

Aethylred commented:
No checks have failed, they were all cancelled/timed-out :(

Aethylred commented:
I followed https://docs.ansible.com/ansible/2.5/dev_guide/developing_rebasing.html and all the flags went away

Aethylred commented:
Yay, it still rebases

Akasurde commented:
Hi @Aethylred, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
